File: Lab 4/C-sharp/MyString.cs
namespace Library
{
    class MyString
    {

        public MyString()
        {
            length = 0;
        }

        public MyString(char[] str)
        {
            length = str.Length;
            this.str = new char[length];
            for (int i = 0; i < length; i++)
            {
                this.str[i] = str[i];
            }
        }

        public MyString(in MyString other)
        {
            length = other.length;
            str = new char[length];
            for (int i = 0; i < length; i++)
            {
                str[i] = other.str[i];
            }
        }

        public int Length
        {
            get
            {
                return length;
            }
        }

        public char[] Line
        {
            get
            {
                return str;
            }
        }

        public static MyString operator + (in MyString left,in MyString right)
        {
            MyString newStr = new MyString();
            newStr.length = left.length + right.length;
            newStr.str = new char[newStr.length];
            int i;
            for (i=0; i < left.length; i++)
            {
                newStr.str[i] = left.str[i];
            }
            for(int j = 0; j < right.length; j++, i++)
            {
                newStr.str[i] = right.str[j];
            }
            return newStr;
        }

        public static MyString operator + (in MyString left, char symbol)
        {
            MyString newStr = new MyString();
            newStr.length = left.length + 1;
            newStr.str = new char[newStr.length];
            int i;
            for (i = 0; i < left.length; i++)
            {
                newStr.str[i] = left.str[i];
            }
            newStr.str[newStr.length] = symbol;
            return newStr;
        }

        public static MyString operator - (in MyString left, char symbol)
        {
            int counter = 0;
            for(int i = 0; i < left.length; i++)
            {
                if (left.str[i] == symbol)
                {
                    counter++;
                }
            }
            MyString newStr = new MyString();
            newStr.length = left.length - counter;
            newStr.str = new char[newStr.length];
            int pos = 0;
            for(int i = 0; i < left.length; i++)
            {
                if (left.str[i] != symbol)
                {
                    newStr.str[pos] = left.str[i];
                    pos++;
                }
            }
            return newStr;
        }

        private char[] str;
        private int length;
    }
}

File: Lab 4/C++/MyString.cpp
#include "MyString.h"

MyString::MyString() {
	str = nullptr;
	length = 0;
}

MyString::MyString(const char* str)
{
	length = 0;
	while (str[length] != '\0')
		length++;
	this->str = new char[length + 1];
	for (int i = 0; i < length; i++) {
		this->str[i] = str[i];
	}
	this->str[length] = '\0';
}

MyString::MyString(const MyString& other) {
	length = other.length;
	str = new char[length+1];
	for (int i = 0; i < length; i++) {
		str[i] = other.str[i];
	}
	str[length] = '\0';
}

MyString& MyString::operator = (const MyString& other) {
	if (str != nullptr) {
		delete[]str;
	}
	length = other.length;
	str = new char[length + 1];
	for (int i = 0; i < length; i++) {
		str[i] = other.str[i];
	}
	str[length] = '\0';
	return *this;
}

MyString::~MyString() {
	delete[]str;
}

int MyString::GetLength() {
	return length;
}

char* MyString::GetLine() {
	return str;
}

MyString MyString::operator +(const MyString& other) {
	MyString newStr;
	int thisLength = this->length;
	int otherLength = other.length;
	newStr.length=thisLength + otherLength;
	newStr.str = new char[newStr.length + 1];
	int i = 0;
	for (i; i < thisLength; i++) {
		newStr.str[i] = this->str[i];
	}
	for(int j=0;j<otherLength;j++,i++){
		newStr.str[i] = other.str[j];
	}
	newStr.str[thisLength + otherLength] = '\0';
	return newStr;
}

MyString MyString::operator +(char symbol) {
	MyString newStr;
	newStr.length = length+1;
	newStr.str = new char[newStr.length + 1];
	for (int i = 0; i < length; i++) {
		newStr.str[i] = str[i];
	}
	newStr.str[length] = symbol;
	newStr.str[newStr.length] = '\0';
	return newStr;
}

MyString MyString::operator -(char symbol) {
	int counter=0;
	for (int i = 0; i < length; i++) {
		if (str[i] == symbol)
			counter++;
	}
	MyString newStr;
	newStr.length = length - counter;
	newStr.str = new char[newStr.length + 1];
	int i = 0;
	for (int j = 0; j < length; j++) {
		if (str[j] != symbol) {
			newStr.str[i] = str[j];
			i++;
		}
	}
	newStr.str[newStr.length] = '\0';
	return newStr;
}